bootstrap/0036_sql_tools.sql

```diff
@@ -0,0 +1,154 @@
+execute as owner
+as
+begin
+    let keys array := object_keys(:obj);
+
+    -- The USING portion of the merge statement, e.g. 'obj['col1'] as col1, ...'
+    let using_select_cols array := (select array_agg('obj[\'' || value || '\'] as ' || value) from table(flatten(input=>:keys)));
+    let using_expr text := (select array_to_string(:using_select_cols, ', '));
+
+    -- The UPDATE portion of the merge statement, e.g. 'dest.col1 = src.col1, ...'
+    let update_cols array := (select array_agg('dest.' || value || ' = src.' || value) from table(flatten(input=>:keys)));
+    let update_expr text := (select array_to_string(:update_cols, ', '));
+
+    -- The INSERT portion of the merge statement
+    let insert_target_expr text := (select array_to_string(:keys, ', '));
+    -- e.g. src.col1, src.col2, ...
+    let insert_query_cols array := (select array_agg('src.' || value) from table(flatten(input=>:keys)));
+    let insert_query_expr text := (select array_to_string(:insert_query_cols, ', '));
+
+    let merge_tmpl varchar := $$MERGE INTO internal.{table} dest USING (
+SELECT {using_expr} FROM (SELECT PARSE_JSON(?) AS obj)) src
+ON dest.{key} = src.{key}
+WHEN MATCHED THEN
+    UPDATE SET {update_expr}
+WHEN NOT MATCHED THEN
+    INSERT ({insert_target_expr}) values ({insert_query_expr})
+$$;
+    let merge_stmt varchar := (select tools.templatejs(:merge_tmpl,
+        {'table': :tbl, 'using_expr': :using_expr, 'key': :key, 'update_expr': :update_expr,
+            'insert_target_expr': :insert_target_expr, 'insert_query_expr': :insert_query_expr}));
+    execute immediate :merge_stmt using (obj);
+end;
+
+CREATE OR REPLACE FUNCTION VALIDATION.IS_NULL(name text)
+    RETURNS text
+    COMMENT='Validates that a field in an object is null'
+AS
+$$
+    'TO_CHAR(f:' || name || ') is null'
+$$;
+
+CREATE OR REPLACE FUNCTION VALIDATION.NOT_NULL(name text)
+    RETURNS text
+    COMMENT='Validates that a field in an object is not null'
+AS
+$$
+    'TO_CHAR(f:' || name || ') is not null'
+$$;
+
+CREATE OR REPLACE FUNCTION VALIDATION.LABEL_ABSENT()
+    RETURNS text
+AS
+$$
+    '(select count(*) = 0 from INTERNAL.LABELS where group_name is null and name = f:name)'
+$$;
+
+CREATE OR REPLACE FUNCTION VALIDATION.GROUPED_LABEL_ABSENT()
+    RETURNS TEXT
+AS
+$$
+    '(select count(*) = 0 from INTERNAL.LABELS where group_name = f:group_name and name = f:name)'
+$$;
+
+CREATE OR REPLACE FUNCTION VALIDATION.GROUP_NAME_UNIQUE()
+    RETURNS TEXT
+AS
+$$
+    '(select count(*) = 0 from internal.labels where name = f:group_name and group_name is null)'
+$$;
+
+CREATE OR REPLACE PROCEDURE VALIDATION.VALID_LABEL_CONDITION(input varchar)
+    RETURNS TABLE(result BOOLEAN)
+    LANGUAGE SQL
+AS
+BEGIN
+    let c varchar := (select parse_json(:input):condition);
+    let stmt varchar := (select 'select ' || :c || ' from reporting.enriched_query_history limit 1');
+    execute immediate :stmt;
+    -- if the condition is invalid, we let the exception propagate.
```

Comment:
why do we allow the exception to propagate here but below we capture it?

Reply:
We don't managed the exceptions below so we can distinguish `statement_error` from `other` exceptions. In this procedure, any exception indicates a failed validation (and we can let the caller handle it since we have the caller doing this as a fail-safe).